src/plonk/prover.rs

```diff
@@ -20,256 +20,399 @@ use crate::transcript::TranscriptWrite;
 pub fn create_proof<C: CurveAffine, T: TranscriptWrite<C>, ConcreteCircuit: Circuit<C::Scalar>>(
     params: &Params<C>,
     pk: &ProvingKey<C>,
-    circuit: &ConcreteCircuit,
-    aux: &[Polynomial<C::Scalar, LagrangeCoeff>],
+    circuits: &[ConcreteCircuit],
+    auxs: &[&[Polynomial<C::Scalar, LagrangeCoeff>]],
     transcript: &mut T,
 ) -> Result<(), Error> {
-    if aux.len() != pk.vk.cs.num_aux_columns {
-        return Err(Error::IncompatibleParams);
-    }
-
-    struct WitnessCollection<F: Field> {
-        advice: Vec<Polynomial<F, LagrangeCoeff>>,
-        _marker: std::marker::PhantomData<F>,
-    }
-
-    impl<F: Field> Assignment<F> for WitnessCollection<F> {
-        fn assign_advice(
-            &mut self,
-            column: Column<Advice>,
-            row: usize,
-            to: impl FnOnce() -> Result<F, Error>,
-        ) -> Result<(), Error> {
-            *self
-                .advice
-                .get_mut(column.index())
-                .and_then(|v| v.get_mut(row))
-                .ok_or(Error::BoundsFailure)? = to()?;
-
-            Ok(())
-        }
-
-        fn assign_fixed(
-            &mut self,
-            _: Column<Fixed>,
-            _: usize,
-            _: impl FnOnce() -> Result<F, Error>,
-        ) -> Result<(), Error> {
-            // We only care about advice columns here
-
-            Ok(())
-        }
-
-        fn copy(&mut self, _: usize, _: usize, _: usize, _: usize, _: usize) -> Result<(), Error> {
-            // We only care about advice columns here
-
-            Ok(())
+    for aux in auxs.iter() {
+        if aux.len() != pk.vk.cs.num_aux_columns {
+            return Err(Error::IncompatibleParams);
         }
     }
 
     let domain = &pk.vk.domain;
     let mut meta = ConstraintSystem::default();
     let config = ConcreteCircuit::configure(&mut meta);
 
-    let mut witness = WitnessCollection {
-        advice: vec![domain.empty_lagrange(); meta.num_advice_columns],
-        _marker: std::marker::PhantomData,
-    };
-
-    // Synthesize the circuit to obtain the witness and other information.
-    circuit.synthesize(&mut witness, config)?;
-
-    let witness = witness;
-
-    // Compute commitments to aux column polynomials
-    let aux_commitments_projective: Vec<_> = aux
-        .iter()
-        .map(|poly| params.commit_lagrange(poly, Blind::default()))
-        .collect();
-    let mut aux_commitments = vec![C::zero(); aux_commitments_projective.len()];
-    C::Projective::batch_to_affine(&aux_commitments_projective, &mut aux_commitments);
-    let aux_commitments = aux_commitments;
-    drop(aux_commitments_projective);
-    metrics::counter!("aux_commitments", aux_commitments.len() as u64);
-
-    for commitment in &aux_commitments {
-        transcript
-            .common_point(*commitment)
-            .map_err(|_| Error::TranscriptError)?;
+    struct AuxSingle<C: CurveAffine> {
+        pub aux_values: Vec<Polynomial<C::Scalar, LagrangeCoeff>>,
```

Comment:
```suggestion
    struct AuxSingle<'a, C: CurveAffine> {
        pub aux_values: &'a [Polynomial<C::Scalar, LagrangeCoeff>],
```
Since we don't mutate this, we can just use the slices that are passed in from the caller.

Reply:
There will be other things affected by the lifetime parameter here which will need to be worked out.